### PR TITLE
CIRCSTORE-89: Patron Notice Policy POST & PUT API Endpoints

### PR DESCRIPTION
**Pull request by opochepam**
1. Create Patron Notice Policy schema
2. Implement Patron Notice Policy POST and PUT endpoints
3. Write POST and PUT API tests
4. Stop embedded postgres in API tests
5. Update gitignore - add *.iml

### CODE REVIEW

File: ramls/patron-notice-policy.json [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+      "type": "string",
+      "description": "Patron notice policy name"
+    },
+    "description": {
+      "type": "string",
+      "description": "Patron notice policy description"
+    },
+    "active": {
+      "type": "boolean",
+      "description": "Patron notice policy description"
```
[the remark this answers is not in the document]
Comment (opochepam) [reply ramls/patron-notice-policy.json]: fixed

File: ramls/patron-notice-policy.json [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+    },
+    "metadata": {
+      "description": "Metadata about creation and changes to requests, provided by the server (client should not provide)",
```
[the remark this answers is not in the document]
Comment (opochepam) [reply ramls/patron-notice-policy.json]: fixed

File: descriptors/ModuleDescriptor-template.json [hunk 3]
```diff
@@ -481,7 +497,9 @@
         "circulation-storage.cancellation-reasons.item.post",
         "circulation-storage.cancellation-reasons.item.put",
         "circulation-storage.cancellation-reasons.collection.delete",
-        "circulation-storage.cancellation-reasons.item.delete"
+        "circulation-storage.cancellation-reasons.item.delete",
+        "circulation-storage.patron-notice-policies.item.create",
```
Comment (marcjohnson-kint) [inline descriptors/ModuleDescriptor-template.json]: This naming is inconsistent with the current convention of using post as the final postfix in the permission name for a POST endpoint. Please can this be changed to fit with the current convention.
```suggestion
        "circulation-storage.patron-notice-policies.item.post",
```

(admittedly this convention is a little inconsistent as it uses item for collection level posts)

File: descriptors/ModuleDescriptor-template.json [hunk 4]
```diff
@@ -481,7 +497,9 @@
...
+        "circulation-storage.patron-notice-policies.item.update"
```
Comment (marcjohnson-kint) [inline descriptors/ModuleDescriptor-template.json]: This naming is inconsistent with the current convention of using put as the final postfix in the permission name for a PUT endpoint. Please can this be changed to fit with the current convention.
```suggestion
        "circulation-storage.patron-notice-policies.item.put"
```

File: descriptors/ModuleDescriptor-template.json [hunk 5]
```diff
@@ -221,6 +221,22 @@
         }
       ]
     },
+    {
+      "id": "patron-notice-policy-storage",
+      "version": "1.0",
```
Comment (marcjohnson-kint) [inline descriptors/ModuleDescriptor-template.json]: Given that this interface is very partial at the moment (as the collection and item GET endpoints have been separated out), I think it is worth considering using a 0.x version here.

For context, previous conversation in FOLIO agreed that an interface version is effectively published when a change is merged to the master branch of the implementing module. This means that the subsequent work (CIRCSTORE-90, CIRCSTORE-91, CIRCSTORE-92 etc all need to change the interface version as well.

Does that make sense?

File: src/main/java/org/folio/rest/impl/PatronNoticePoliciesAPI.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        asyncResultHandler.handle(Future.succeededFuture(
+          PutPatronNoticePolicyStoragePatronNoticePoliciesByPatronNoticePolicyIdResponse.respond500WithTextPlain(e)));
+      }
+    });
+  }
+
+  private Errors createNotUniqueNameErrors(String name) {
+    Error error = new Error();
+    error.setMessage(String.format("'%s' name in not unique", name));
+    error.setCode(STATUS_CODE_DUPLICATE_NAME);
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/rest/impl/PatronNoticePoliciesAPI.java]: Please see [comment on the issue](https://issues.folio.org/browse/CIRCSTORE-89?focusedCommentId=42707&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-42707) regarding the use of code for error identification

File: descriptors/ModuleDescriptor-template.json (hunk 3 above)
Comment (opochepam) [reply descriptors/ModuleDescriptor-template.json]: changed to post

Comment (opochepam) [reply descriptors/ModuleDescriptor-template.json]: actually it's item level post, there is no possibility to post a collection

File: descriptors/ModuleDescriptor-template.json (hunk 4 above)
Comment (opochepam) [reply descriptors/ModuleDescriptor-template.json]: changed to put

File: descriptors/ModuleDescriptor-template.json [hunk 7]
```diff
@@ -221,6 +221,22 @@
...
       ]
     },
+    {
+      "id": "patron-notice-policy-storage",
+      "version": "0.1",
+      "handlers": [
+        {
+          "methods": ["POST"],
+          "pathPattern": "/patron-notice-policy-storage/patron-notice-policies",
+          "permissionsRequired": ["circulation-storage.patron-notice-policies.item.create"]
```
Comment (marcjohnson-kint) [inline descriptors/ModuleDescriptor-template.json]: This also needs updating to reflect the renamed permission
```suggestion
          "permissionsRequired": ["circulation-storage.patron-notice-policies.item.post"]
```

File: descriptors/ModuleDescriptor-template.json [hunk 8]
```diff
@@ -221,6 +221,22 @@
...
+        },
+        {
+          "methods": ["PUT"],
+          "pathPattern": "/patron-notice-policy-storage/patron-notice-policies/{id}",
+          "permissionsRequired": ["circulation-storage.patron-notice-policies.item.update"]
```
Comment (marcjohnson-kint) [inline descriptors/ModuleDescriptor-template.json]: This also needs updating to reflect the renamed permission
```suggestion
          "permissionsRequired": ["circulation-storage.patron-notice-policies.item.put"]
```

File: descriptors/ModuleDescriptor-template.json (hunk 7 above)
Comment (opochepam) [reply descriptors/ModuleDescriptor-template.json]: thanks, it makes sense

Comment (opochepam) [reply descriptors/ModuleDescriptor-template.json]: fixed

File: descriptors/ModuleDescriptor-template.json (hunk 8 above)
Comment (opochepam) [reply descriptors/ModuleDescriptor-template.json]: fixed

**Pull request merged**